pkg/asset/manifests/template.go

```diff
@@ -6,6 +6,17 @@ type AwsCredsSecretData struct {
 	Base64encodeSecretAccessKey string
 }
 
+// AzureCredsSecretData holds encoded credentials and is used to generate cloud-creds secret
+type AzureCredsSecretData struct {
```

Comment:
nit: `Base64encodeResourcePrefix`  `Base64encodeResourceGroup` `Base64encodeRegion` do not seem like they belong to credentials secret....

Reply:
That's true, but they are required for cluster-api-provider-azure, so that's the easiest solution for now.

Reply:
> That's true, but they are required for cluster-api-provider-azure, so that's the easiest solution for now.

can you expand how each one is used. all these should be part of the cluster-api-provider definition....

Reply:
Each field here is part of cluster specific information, technically this could be renamed to clusterresources, this cannot be part of machine spec since machine spec would only define behaviour of the machine or not cluster specific information. 

ResourcePrefix, ResourceGroup and Region is used to get the information for base resources like VNET, Subnet, LoadBalancer etc, for which we need the credentials stored here.

We could create another resource here apart from azure credentials, but would be overload of configurations in the cluster.

Reply:
> Each field here is part of cluster specific information, technically this could be renamed to clusterresources, this cannot be part of machine spec since machine spec would only define behaviour of the machine or not cluster specific information.
> 
> ResourcePrefix, ResourceGroup and Region is used to get the information for base resources like VNET, Subnet, LoadBalancer etc, for which we need the credentials stored here.
> 
> We could create another resource here apart from azure credentials, but would be overload of configurations in the cluster.

> That's true, but they are required for cluster-api-provider-azure, so that's the easiest solution for now.

Now i'm confused, there are part of Machine definition, but they are required for cluster-api-provider-azure how is that possible?

> ResourcePrefix, ResourceGroup and Region is used to get the information for base resources like VNET, Subnet, LoadBalancer etc, for which we need the credentials stored here.

why is the cluster-api-provider-azure guessing these, we want to place the machines in installer defined locations.

Reply:
There are cluster specific resources(aka base resources) and machine specific resources, cluster specific resources include VNET, subnets, load balancers etc. Machine specific resources include, VMSize, Userdata (for bootstrapping), etc. Machine though it uses base resources for reference to build on top of, this is pretty much static and can be referred from cluster specification (which is the current behaviour of cluster api provider azure https://github.com/kubernetes-sigs/cluster-api-provider-azure). Since we dont have a cluster actuator in openshift, we need to construct the names of the resources for which it can be referred by using `ResourcePrefix`, `ResourceName` and `Region`, it doesnt make sense to be part of machine spec since its not a configurable resource. There are 2 options

1. Move these configuration options and provide a separate configmap which can be referred in machine actuator (all the dependent resource names)
2. Keep the current implementation and rename the `azure-credentials` to `azure-clusterconfig`

We are open for either, though first option is cleaner, just adds a overhead of managing yet another configmap input to machine spec

Reply:
Here is the usage, of this configuration 

https://github.com/openshift/cluster-api-provider-azure/blob/f9a5d431c6f7b4d21b5079cd09a5359e75961667/pkg/cloud/azure/actuators/machine_scope.go#L283

How resource names are constructed!

https://github.com/openshift/cluster-api-provider-azure/blob/master/pkg/cloud/azure/defaults.go

Reply:
if this is the state of the art currently, let's keep it as-is.